internal/pkg/gateway/commit/notifier.go

```diff
@@ -0,0 +1,72 @@
+/*
+Copyright 2021 IBM All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package commit
+
+import (
+	"context"
+	"sync"
+
+	"github.com/hyperledger/fabric-protos-go/peer"
+	"github.com/hyperledger/fabric/core/ledger"
+)
+
+type LedgerNotifier interface {
+	CommitNotifications(done <-chan struct{}, channelName string) (<-chan *ledger.CommitNotification, error)
+}
+
+type Notifier struct {
+	lock             sync.Mutex
+	ledgerNotifier   LedgerNotifier
+	channelNotifiers map[string]*channelNotifier
+}
+
+type Notification struct {
+	BlockNumber    uint64
+	TransactionID  string
+	ValidationCode peer.TxValidationCode
+}
+
+func NewNotifier(ledgerNotifier LedgerNotifier) *Notifier {
+	if ledgerNotifier == nil {
+		panic("nil ledger notifier")
+	}
```

Comment:
this shouldn't be necessary as this isn't library code

Reply:
I prefer failing fast to debugging obscure errors further down the line if the code is accidentally mis-used. Is it a significant issue to leave this in place, or would it be better to change `NewNotifier` to return `(*Notifier, error)` and force all callers to error check even if they are passing valid arguments that will never generate the error?

Reply:
This is not a library. In fact, there's no need for a constructor here at all if the field were embedded or exported. You're choosing to implement a constructor.

Now, do i care that you have a nil check and a panic? Not really. Do I care that it implies we don't trust the caller? Yes.

Reply:
You have previously instructed me to follow the guidance in Effective Go. This states:

> it's helpful to arrange when designing your data structures that the zero value of each type can be used without further initialization

and then goes on to say:

> Sometimes the zero value isn't good enough and an initializing constructor is necessary

For this type the zero value is not usable and will generate a panic if used. Usable instances are non-trivial for calling code to construct correctly and would also couple callers to implementation detail of this type that they don’t need to know or care about.

You might be inferring that I don’t trust calling code but I am absolutely not implying that. I am expecting to be writing the calling code and I want to make it as easy as possible for me to get that code right.

I am trying really really hard to follow all of your directives but it seems to be impossible to know when there is some other directive I haven’t been told about that supersedes your earlier directives.

Reply:
I have changed it to return nil rather than panic to align exactly with the examples in Effective Go